File: aula02-pilhas-filas/c/central_impressao.c
#include <stdio.h>
#include <stdlib.h>
#include "fila.h"
#include "pilha.h"

static void exigir(int condicao, const char *mensagem) {
    if (!condicao) {
        fprintf(stderr, "Erro: %s\n", mensagem);
        exit(EXIT_FAILURE);
    }
}

static int processar_proximo(Fila *espera, Pilha *historico) {
    int trabalho;
    if (!fila_desenfileirar(espera, &trabalho)) {
        return 0;
    }
    exigir(pilha_empilhar(historico, trabalho), "historico cheio");
    printf("Trabalho processado: %d\n", trabalho);
    return 1;
}

int main(void) {
    Fila espera;
    Pilha historico;
    int valor;

    fila_criar(&espera);
    pilha_criar(&historico);

    exigir(fila_enfileirar(&espera, 101), "nao foi possivel criar o trabalho 101");
    exigir(fila_enfileirar(&espera, 102), "nao foi possivel criar o trabalho 102");
    exigir(fila_enfileirar(&espera, 103), "nao foi possivel criar o trabalho 103");

    exigir(processar_proximo(&espera, &historico), "fila vazia");
    exigir(processar_proximo(&espera, &historico), "fila vazia");
    exigir(fila_enfileirar(&espera, 104), "nao foi possivel criar o trabalho 104");

    printf("Fila de espera: ");
    fila_exibir(&espera);
    printf("Historico - topo para base: ");
    pilha_exibir(&historico);

    exigir(fila_consultar_frente(&espera, &valor), "fila vazia");
    printf("Proximo trabalho: %d\n", valor);
    exigir(pilha_consultar_topo(&historico, &valor), "historico vazio");
    printf("Ultimo trabalho processado: %d\n", valor);

    fila_destruir(&espera);
    return 0;
}

File: aula02-pilhas-filas/python/central_impressao.py
from fila import Fila
from pilha import Pilha


def processar_proximo(espera, historico):
    if espera.esta_vazia():
        return False
    trabalho = espera.desenfileirar()
    historico.empilhar(trabalho)
    print(f"Trabalho processado: {trabalho}")
    return True


espera = Fila()
historico = Pilha()

for trabalho in (101, 102, 103):
    espera.enfileirar(trabalho)

processar_proximo(espera, historico)
processar_proximo(espera, historico)
espera.enfileirar(104)

print("Fila de espera:", *espera.valores())
print("Historico - topo para base:", *historico.valores_do_topo_para_base())
print("Proximo trabalho:", espera.consultar_frente())
print("Ultimo trabalho processado:", historico.consultar_topo())
